### ISSUE TITLE
Custom Eloquent builder methods returning self are reported as PossiblyUnusedReturnValue when call sites discard the return

### ISSUE THREAD

**Issue by lptn** [psalm/laravel-psalm-plugin#1448]
**Describe the bug**

A method on a custom Eloquent builder that returns `self`/`static` is reported as `PossiblyUnusedReturnValue` whenever every call site discards the return value.

Discarding it is correct Laravel. `Illuminate\Database\Eloquent\Builder` is mutable: `where()`, `whereNull()` and friends mutate `$this->query` and return `$this` only as a convenience. So `$query->published();` and `$query = $query->published();` are equivalent, and the statement form is the idiomatic one when a builder is handed to you to modify in place — a `Closure` filter callback, a policy scope, a `Scope` object, a shared query-modifier helper.

Psalm has an exemption for this shape (`$method_storage->probably_fluent`, checked in `Psalm\Internal\Codebase\ClassLikes::checkMethodReferences()`), but its detection is purely syntactic (`Psalm\Internal\PhpVisitor\Reflector\FunctionLikeNodeScanner`):

```php
$last_stmt = end($stmt->stmts);

if ($last_stmt instanceof PhpParser\Node\Stmt\Return_
    && $last_stmt->expr instanceof PhpParser\Node\Expr\Variable
    && $last_stmt->expr->name === 'this'
) {
    $storage->probably_fluent = true;
}
```

It requires a literal `return $this;`. A builder method is almost never written that way — it is written `return $this->whereNull(...);`, which is the same thing through Eloquent's own fluent API. The declared return type is not consulted at all, so `self` vs `static` makes no difference.

Psalm cannot fix this generically: for an immutable fluent API, discarding the return really is a bug, and Psalm has no way to know which is which. This plugin does know. It already owns custom-builder semantics (`CustomBuilderMethodHandler`, registered per builder class by `ModelRegistrationHandler`), so it can tell that a method declared on an `Illuminate\Database\Eloquent\Builder` subclass and returning `self`/`static`/the builder type is mutating-fluent by construction.

The failure is also unstable in a way that makes it read as noise rather than a finding. On a real codebase, a builder with ten such methods reported exactly two: the two whose *only* call site was a statement. The eight others had at least one chained call site somewhere (`->published()->get()`), which registers a return reference and clears the issue for every other call site too. Same code shape, same correctness, opposite verdict, decided by an unrelated file.

**Code to reproduce**

```php
// app/PsalmRepro/Post.php
namespace App\PsalmRepro;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Query\Builder as QueryBuilder;

final class Post extends Model
{
    /** @param QueryBuilder $query */
    #[\Override]
    public function newEloquentBuilder($query): PostBuilder
    {
        return new PostBuilder($query);
    }
}
```

```php
// app/PsalmRepro/PostBuilder.php
namespace App\PsalmRepro;

use Illuminate\Database\Eloquent\Builder;

/** @extends Builder<Post> */
final class PostBuilder extends Builder
{
    public function published(): self
    {
        return $this->whereNotNull('published_at');
    }

    public function draft(): self
    {
        return $this->whereNull('published_at');
    }
}
```

```php
// app/PsalmRepro/PostFilter.php
namespace App\PsalmRepro;

final class PostFilter
{
    public function apply(PostBuilder $query, bool $onlyPublished): void
    {
        if ($onlyPublished) {
            $query->published();
        } else {
            $query->draft();
        }
    }
}
```

```php
// app/PsalmRepro/registration.php
// Class references only; without one Psalm reports UnusedClass and never inspects members.
return [
    \App\PsalmRepro\Post::class,
    \App\PsalmRepro\PostBuilder::class,
    \App\PsalmRepro\PostFilter::class,
];
```

Full-project run, `findUnusedCode="true"`:

```text
PossiblyUnusedReturnValue  app/PsalmRepro/PostBuilder.php:12  The return value for this method is never used
PossiblyUnusedReturnValue  app/PsalmRepro/PostBuilder.php:17  The return value for this method is never used
```

Expected: no findings. Add `->get()` after either call and that method's finding disappears while the other remains, which shows the verdict is driven by an unrelated call site rather than by the method.

**Impacted Versions**

```text
laravel/framework           v13.29.0
psalm/plugin-laravel        dev-master
psalm/psalm-plugin-api      0.1.0
vimeo/psalm                 7.0.0-beta19
```

PHP 8.5.

**Psalm config**

```xml
<?xml version="1.0"?>
<psalm
        errorLevel="1"
        resolveFromConfigFile="true"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns="https://getpsalm.org/schema/config"
        xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
        findUnusedCode="true"
>
    <projectFiles>
        <directory name="app"/>
        <ignoreFiles>
            <directory name="vendor"/>
        </ignoreFiles>
    </projectFiles>
    <issueHandlers>
        <UnusedClass errorLevel="info"/>
        <PossiblyUnusedMethod errorLevel="info"/>
    </issueHandlers>
    <plugins>
        <pluginClass class="Psalm\LaravelPlugin\Plugin">
            <experimental value="true"/>
            <failOnInternalError value="true"/>
        </pluginClass>
    </plugins>
</psalm>
```

**Additional context**

Possible fix: in an `AfterClassLikeVisit` handler, for a class extending `Illuminate\Database\Eloquent\Builder`, set `$methodStorage->probably_fluent = true` on public methods whose declared return type is `self`, `static`, `$this` or the builder class itself. `checkMethodReferences()` reads that flag directly (`$method_storage->is_static || !$method_storage->probably_fluent`), so nothing else needs to change, and methods returning a model, a collection, a count or `void` keep reporting normally.

Scope question worth deciding: whether the same treatment should extend to legacy `scopeXxx()` and `#[Scope]` methods that return a builder, and to `Illuminate\Database\Query\Builder` subclasses. Both are the same mutable-fluent contract, though scopes are reached through `__call` and may already be covered by other reference edges.

A narrower alternative is to only exempt methods whose body ends in `return $this->...;` on the builder, which is closer to Psalm's existing heuristic, but it would still miss a method ending in a conditional return and does not add much over the return-type check.
